bot/exts/utils/snekbox/_eval.py

```diff
@@ -0,0 +1,129 @@
+            args=[path],
+            files=[FileAttachment(path, code.encode())],
+        )
+
+    def as_version(self, version: PythonVersion) -> EvalJob:
+        """Return a copy of the job with a different Python version."""
+        return EvalJob(
+            args=self.args,
+            files=self.files,
+            name=self.name,
+            version=version,
+        )
+
+    def to_dict(self) -> dict[str, list[str | dict[str, str]]]:
+        """Convert the job to a dict."""
+        return {
+            "args": self.args,
+            "files": [file.to_dict() for file in self.files],
+        }
+
+
+@dataclass(frozen=True)
+class EvalResult:
+    """The result of an eval job."""
+
+    stdout: str
+    returncode: int | None
+    files: list[FileAttachment] = field(default_factory=list)
+    failed_files: list[str] = field(default_factory=list)
+
+    @property
+    def status_emoji(self) -> str:
+        """Return an emoji corresponding to the status code or lack of output in result."""
+        # If there are attachments, skip empty output warning
+        if not self.stdout.strip() and not self.files:  # No output
+            return ":warning:"
+        elif self.returncode == 0:  # No error
+            return ":white_check_mark:"
+        else:  # Exception
+            return ":x:"
+
+    def get_message(self, job: EvalJob) -> tuple[str, str]:
+        """Return a user-friendly message and error corresponding to the process's return code."""
+        msg = f"Your {job.version} {job.name} job has completed with return code {self.returncode}"
+        error = ""
+
+        if self.returncode is None:
+            msg = f"Your {job.version} {job.name} job has failed"
+            error = self.stdout.strip()
+        elif self.returncode == 128 + SIGKILL:
+            msg = f"Your {job.version} {job.name} job timed out or ran out of memory"
+        elif self.returncode == 255:
+            msg = f"Your {job.version} {job.name} job has failed"
+            error = "A fatal NsJail error occurred"
+        else:
+            # Try to append signal's name if one exists
+            with contextlib.suppress(ValueError):
+                name = Signals(self.returncode - 128).name
+                msg = f"{msg} ({name})"
+
+        # Add error message for failed attachments
+        if self.failed_files:
+            failed_files = f"({', '.join(self.failed_files)})"
+            # Case for over 10
+            if len(self.failed_files) + len(self.files) > FILE_COUNT_LIMIT:
+                msg += (
+                    f".\n\n> Some files were not able to be uploaded, as they exceeded"
+                    f" the {FILE_COUNT_LIMIT} file upload limit {failed_files}"
+                )
+            else:
+                msg += (
+                    f".\n\n> Some files were not able to be uploaded {failed_files}."
+                    f" Check that the file size is less than {sizeof_fmt(FILE_SIZE_LIMIT)}"
+                )
```

Comment:
I'm not a fan of how this looks when inside a quote block. I also think this should be moved below the code block showing the output, which feels more consistent with other messages that conditionally appear. Any strong opinions against that?

To facilitate that change, it would be good to create separate functions to get all these different strings instead of returning a tuple.

Reply:
Separate functions implemented, and added some unit tests since [d2de465](https://github.com/python-discord/bot/pull/2326/commits/d2de465e8fb3659eb1fe40aa2d1c9e9cb80e0d11). Images in other review comment thread https://github.com/python-discord/bot/pull/2326/files/56664828c300b0380e16cde66f6106b0fdcf56bd..d00584408a20e231df1e3d6ce3c4837e6bb829e4#diff-578c89688caf1d45de35421fc8343ffa1d3e675aeda44a6d476d11a3ff286322

Reply:
Thanks for refactoring that. Returning a tuple there has been bugging me for a while now.